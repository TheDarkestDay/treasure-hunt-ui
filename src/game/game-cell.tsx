import React from 'react';
import { RevealedCell, Point } from '../domain';
import styles from './game-cell.module.css';

export interface GameCellProps {
    x: number;
    y: number;
    revealedCells: RevealedCell[];
    cellsToReveal: Point[];
    onClick: (i: number, j: number) => void;
}

export const GameCell = ({revealedCells, x, y, onClick, cellsToReveal}: GameCellProps) => {
    const cellContent = revealedCells.find((cell) => cell.x === x && cell.y === y)?.content;
    const isCellAboutToBeRevealed = cellsToReveal.some((cell) => cell.x === x && cell.y === y);

    let cellClass = styles.gameFieldCell;
    if (isCellAboutToBeRevealed) {
        cellClass += ` ${styles.gameFieldCellToReveal}`
    }

    const handleClick = () => {
        onClick(x, y);
    };

    return (
        <div className={cellClass} onClick={handleClick}>
            {cellContent || ''}
        </div>
    );
};