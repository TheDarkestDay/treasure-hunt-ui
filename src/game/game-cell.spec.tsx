import React from 'react';
import { mount } from 'enzyme';
import { GameCell } from './game-cell';

describe('GameCell', () => {
    it('should match snapshot without highlight', () => {
        const gameCell = mount(<GameCell x={0} y={0} onClick={jest.fn()} revealedCells={[]} cellsToReveal={[]}/>);

        expect(gameCell).toMatchSnapshot();
    });

    it('should match snapshot with highlight', () => {
        const gameCell = mount(<GameCell x={0} y={0} onClick={jest.fn()} revealedCells={[]} cellsToReveal={[{x: 0, y: 0}]}/>);

        expect(gameCell).toMatchSnapshot();
    });

    it('should display the cell content', () => {
        const gameCell = mount(<GameCell x={0} y={0} onClick={jest.fn()} revealedCells={[{x: 0, y: 0, content: 'T'}]} cellsToReveal={[]}/>);

        expect(gameCell).toMatchSnapshot();
    });

    it('should display placeholder if content is absent', () => {
        const gameCell = mount(<GameCell x={0} y={0} onClick={jest.fn()} revealedCells={[{x: 1, y: 2, content: 'T'}]} cellsToReveal={[]}/>);

        expect(gameCell).toMatchSnapshot();
    });

    it('should call the callback with given coordinates', () => {
        const clickHandler = jest.fn();
        const gameCell = mount(<GameCell x={0} y={0} onClick={clickHandler} revealedCells={[{x: 1, y: 2, content: 'T'}]} cellsToReveal={[]}/>);

        gameCell.simulate('click');

        expect(clickHandler).toHaveBeenCalledTimes(1);
        expect(clickHandler).toHaveBeenCalledWith(0, 0);
    });
});