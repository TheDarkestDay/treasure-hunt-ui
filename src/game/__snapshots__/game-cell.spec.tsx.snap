// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCell should display placeholder if content is absent 1`] = `
<GameCell
  cellsToReveal={Array []}
  onClick={[MockFunction]}
  revealedCells={
    Array [
      Object {
        "content": "T",
        "x": 1,
        "y": 2,
      },
    ]
  }
  x={0}
  y={0}
>
  <div
    className="gameFieldCell"
    onClick={[Function]}
  >
    ?
  </div>
</GameCell>
`;

exports[`GameCell should display the cell content 1`] = `
<GameCell
  cellsToReveal={Array []}
  onClick={[MockFunction]}
  revealedCells={
    Array [
      Object {
        "content": "T",
        "x": 0,
        "y": 0,
      },
    ]
  }
  x={0}
  y={0}
>
  <div
    className="gameFieldCell"
    onClick={[Function]}
  >
    T
  </div>
</GameCell>
`;

exports[`GameCell should match snapshot with highlight 1`] = `
<GameCell
  cellsToReveal={
    Array [
      Object {
        "x": 0,
        "y": 0,
      },
    ]
  }
  onClick={[MockFunction]}
  revealedCells={Array []}
  x={0}
  y={0}
>
  <div
    className="gameFieldCell gameFieldCellToReveal"
    onClick={[Function]}
  >
    ?
  </div>
</GameCell>
`;

exports[`GameCell should match snapshot without highlight 1`] = `
<GameCell
  cellsToReveal={Array []}
  onClick={[MockFunction]}
  revealedCells={Array []}
  x={0}
  y={0}
>
  <div
    className="gameFieldCell"
    onClick={[Function]}
  >
    ?
  </div>
</GameCell>
`;
